table_foosball/table_foosball_01.rb

```diff
@@ -0,0 +1,141 @@
+require "securerandom"
+
+class Player
+  def initialize name
+    @name = name
+    @games = [] # what are all the games this player has been involved in?
+    
+    # IDEAS for attributes
+    # - Time stamp for creation of this player object
+    # - UUID because multiple players might have similar names
+    @timestamp = Time.now
+    @id = ID.get_id
+  end
+
+  attr_reader :name, :games, :timestamp, :id
+
+  # return all the games this player won in
+  def won_games
+    @games.select {|g| g.winner.include? self}
+  end
+
+  # return all the games this player lost in
+  def lost_games
+    @games.select {|g| g.loser.include? self}
+  end
+  
+  # return all the games this player has played against opponent o
+  def no_of_games_against o
+    @games.select { |g| g.get_opponents(self).include? o }.length
+  end
+
+  # __The following applies to all of the following methods__
+  # Problem, tho. Returns the first key to hold the max value; ignores rest if same value as max
+  # Return table of player objects/IDs instead, or maybe just false or nil?
+
+  # return the opponent this player has played the most games against
+  def most_games_against
+    most_freq_opp @games
+  end
+
+  # return the opponent this player has most frequently won against
+  def most_wins_against
+    most_freq_opp won_games
+  end
+
+  # most frequently lost against
+  def most_losses_against
+    most_freq_opp lost_games
+  end
+  ####
+
+  private # methods below this point are private
+
+  def most_freq_opp arr
```

Comment:
You could do this simpler by using `#group_by`. This is a super handy method available in pretty much all programming languages.

 ```ruby
all_opponents = games.map { |game| game.opponents(self) }.flatten
all_opponents.group_by { |player| player.name }.max_by { |player, games| games.count }.first
```